### DISCUSSION TITLE
Does the Raspberry Pi Pico (RP2040) PIO USB work in device mode?

### DISCUSSION THREAD

**Discussion by rppicomidi** [hathach/tinyusb#2754] [Q&A] [answered]
@hathach Have you ever tested the PIO USB code for USB Device mode (`dcd_pio_usb.c/.h`)? I am unable to make it work. Details below. Any suggestions you may have would be appreciated.

I am trying to build a USB device that will let me inject data toggle errors. I thought the Pico PIO USB might be a good way to do this because I have access to the low level bus behavior in software. I have tried to build the `midi_test` example with `#define CFG_TUD_RPI_PIO_USB 1` in my `tusb_config.h` file. The code did not build. I then fixed issues with configuration and stubbed out `dcd_sof_enable()` so the code would build. Patches are attached. However, the code still does not work. I am using version 0.6.0 of Pico PIO USB and the latest version of tinyUSB with the attached patches applied. 
[patches_for_dcd_pio_usb.zip](https://github.com/user-attachments/files/16537171/patches_for_dcd_pio_usb.zip)

The beginning of the trace I get using [usb-sniffer-lite](https://github.com/ataradov/usb-sniffer-lite) shows the host repeatedly trying to enumerate the device, but the device does not respond.
```
     0 : --- RESET ---                                                          
   ... : Folded 499 frames                                                      
  1000 : SOF #1933                                                              
    41 : SETUP: 0x00/0                                                          
    44 : DATA0 (8): 80 06 00 01 00 00 08 00                                     
    55 : SETUP: 0x00/0                                                          
    57 : DATA0 (8): 80 06 00 01 00 00 08 00                                     
    68 : SETUP: 0x00/0                                                          
    70 : DATA0 (8): 80 06 00 01 00 00 08 00                                     
    ---The whole trace repeats this pattern forever---
```

**Comment by hathach** [comment, marked as answer]
it should be tested when adding pio_usb, but lots of changes and I haven't tested it since then. device with pio is much harder to work with, since we have to response within the strict timing. While in host we are the one controlling the timing (SOF, token in/out). There is no wonder if that didn't work right away. 

I have no time to fix this, since it has limited usage for me. Maybe you can go back to code (and pio-usb commit) where PR is merged to get it running (but it may still have some hippcup though)

**Reply by rppicomidi** [reply]
Thank you for your quick response. I understand time and priority constraints. I will research this

**Comment by rppicomidi** [comment]
Further notes: One problem is that 0.6.0 release of Pico PIO USB. The `usb_device` example in the Pico PIO USB project does not work for me for Pico PIO USB version 0.6.0. However, with the same target hardware, the `usb_device` example works fine for me if I revert to version 0.5.3. Reverting the Pico PIO USB version to 0.5.3 did not allow TinyUSB `midi_test` example to work with the latest TinyUSB, however, so something else is going on.

**Reply by hathach** [reply]
pio 0.6.0 refractor/rework the pio usage from 2 pio to 1 https://github.com/sekigon-gonnoc/Pico-PIO-USB/pull/120. It also break my arduino support for host as well https://github.com/adafruit/Adafruit_TinyUSB_Arduino/issues/438. Walkaround is change optimization to O0, therefore hint is race/timing condition somewhere. I haven't got time to look into this as well.

**Comment by hathach** [comment]
@rppicomidi I possibly found the root cause  for 0.6.0 -Os issue, https://github.com/sekigon-gonnoc/Pico-PIO-USB/pull/131

**Comment by rppicomidi** [comment]
Thank you, @hathach! Nice job finding it.
